### ISSUE TITLE
Backoffice Product List table UI uplift

### ISSUE THREAD

**Issue by mariocarabotta** [openfoodfoundation/openfoodnetwork#7198]
### What is the problem we are solving
As part of the network inception work, it's been highlighted several times that there's an internal interest to move away from the Inventory functionality altogether. It’s hard to understand for some users, it has been adopted from others when it’s not actually needed, and it also present some technical challenges that we want to minimize in the future.

We’ve also run interviews with producers and hub contributors, asking open ended questions and showing some digital prototypes. Some of the positive feedback we got was around a cleaner user interface for the product list, and consolidating the functionality available in the product list with what’s in inventory. 

From a tech stack point of view, there have been several discussions around moving to React for the backoffice. This is a good opportunity to start the transition.

**Slack channel**
#backoffice-ui-uplift

**Discourse**
**MAIN** Backoffice Product List table uplift: https://community.openfoodnetwork.org/t/backoffice-product-list-table-user-interface-uplift/2098
Network Inception: https://community.openfoodnetwork.org/t/backoffice-product-list-table-user-interface-uplift/2098
A new admin interface for OFN: https://community.openfoodnetwork.org/t/a-new-admin-interface-for-ofn/1784

### Objectives and key metrics

**Product**: Reduce the number of people that use inventory just for ease of use
**Experience**: Provide a better interface for all of OFN enterprises
**Technology**: Move to a more modern system for the backoffice that might also attract new contributors
**Design**: Build the foundations for a backoffice design system

We would measure the success of these objectives by looking at
- reduction in % of variants with variant overrides with at least one order in the last 6 months that don’t have neither stock nor price overwritten
- reduction in % of users using inventory altogether

### How this could look like
Explanatory videos have been posted in the main Discourse thread for community feedback. The main prototypes are:

**[Create first product](https://www.figma.com/proto/kBvKYXnaNU0thSnKJUoyup/Product-List-UI-uplift?node-id=10%3A0&viewport=1492%2C435%2C0.14905129373073578&scaling=min-zoom)**
- Landing experience for a new producer opening the product section of the backoffice
- Creating the first product (no changes)
- Viewing the product list table for the first time

**[Columns selection and table scrolling](https://www.figma.com/proto/kBvKYXnaNU0thSnKJUoyup/Product-List-UI-uplift?node-id=10%3A7088&viewport=116%2C270%2C0.04911578819155693&scaling=scale-down-width&hide-ui=1)**
- Selecting columns
- Scrolling laterally when many columns are visible
- Scrolling vertically

**[Edit fields inline and add tags](https://www.figma.com/proto/kBvKYXnaNU0thSnKJUoyup/Product-List-UI-uplift?node-id=13%3A9644&viewport=90%2C411%2C0.06000276654958725&scaling=scale-down-width&hide-ui=1)**
- Edit unit
- Edit stock
- Add tags to product
- Add tags to variant

**[Filter by categories and tags, sort](https://www.figma.com/proto/kBvKYXnaNU0thSnKJUoyup/Product-List-UI-uplift?node-id=20%3A11&viewport=298%2C-327%2C0.10976928472518921&scaling=min-zoom&hide-ui=1)**
- Filter by category
- Filter by tags
- Tag filter rules
- Sort by stock level

**[On demand and unit display fields](https://www.figma.com/proto/kBvKYXnaNU0thSnKJUoyup/Product-List-UI-uplift?node-id=140%3A12929&viewport=-297%2C243%2C0.10365254431962967&scaling=min-zoom)**
- Change variant to on demand
- Change unit display

### Figma files
Project specific: https://www.figma.com/file/kBvKYXnaNU0thSnKJUoyup/Product-List-UI-uplift
Library: https://www.figma.com/file/A1RDmHC9dMxhjF7JTtiMuB/Components-library

Note: these mockups are based on the [Material-UI](https://material-ui.com/) react library. Icons are part of the [Community Material Design Icons](https://materialdesignicons.com/).

### Milestones
This project is divided into 2 main milestones:
1. Product list table feature parity
2. Table uplift

The first milestone replicates all of the existing functionality of the product list.
The second milestone introduces a few new features that ideally would encourage the users above to migrate from inventory back to product list, specifically:
- ability to add tags to products and variants
- ability to filter product list by tags
- ability to sort table by fields different from creation date (current)

After milestone 1, the improvements could potentially be released to the public, hence splitting this into 2 chunks of work.

**Comment by Matt-Yorkley** [comment]
> CSS Framework: Use of UI component framework vs utility-based framework?

There's no clear plans on this yet, right?

**Comment by Erioldoesdesign** [comment]
> > CSS Framework: Use of UI component framework vs utility-based framework?
> 
> There's no clear plans on this yet, right?

Hmm I think @jibees would be the best person to answer that, we're building a 'pattern library' and I assume that it's compatible with Reactive Rails but unsure if that means we can use Tailwind CSS et.c or not.

Some of what was designed for UI Uplift followed some of the Material UI which is something I'd like to understand more as far as potential restrictions or not

**Comment by Matt-Yorkley** [comment]
I think we need a bit of discussion and a really clear picture of exactly what we're doing in regards to those tech choices before we move this into dev ready :+1:

**Comment by jibees** [comment]
@Matt-Yorkley  What's the problem with Tailwind or any utility first CSS framework? I would be in favor of going on this (and not something like MaterialUI or Bootstrap which is good to create a POC and bootstrap, but a hell to customize as it's not designed for this)

**Comment by Matt-Yorkley** [comment]
The creator of Tailwind themselves says: "If you can suppress the urge to retch long enough to give it a chance, I really think you'll wonder how you ever worked with CSS any other way". I guess I'm still in the retching phase :joy: 

Don't get me wrong, I'm all for utility-based CSS classes, composition over sub-components, and content-agnostic component naming. I'm also not a fan of the BEM approach to solving the traditional problems with inheritance in CSS. 

I'm a big fan of Bulma, mainly because the utility classes are actually _human-readable_.

Bootstrap 4 is not great, but Bootstrap 5 has basically been heavily rewritten to focus on utility classes. MDBootstrap combines Bootstrap 5 and Material Design 2.0 (if that's the aesthetic we're going for?).

If we go with Tailwind we'd be styling every single element from scratch, right? As opposed to having 99% of the styling handled already by a library that already has the look we want?

I don't like the overall results of Tailwind's approach of condensing absolutely _everything_ into tiny utility classes and applying all styling by adding an ever-increasing list of non-human-readable classes-with-modifiers onto each element. eg:

![tailwind-markup](https://user-images.githubusercontent.com/9029026/113324140-0d346300-9317-11eb-9573-127a1b9007c5.png)

Dumping 100% of the styling into the markup doesn't seem like a big leap forward to me... :man_shrugging:

**Comment by jibees** [comment]
> If we go with Tailwind we'd be styling every single element from scratch, right? As opposed to having 99% of the styling handled already by a library that already has the look we want?

I think it's the main point here, driving all the decision we should take in a near future concerning any CSS _framework_.
I personally think that believing in: _99% of the styling is already handled by a library_ is mostly true for a POC driven by developers, but, over time, mostly false for an application driven by a UX/UI team. 

> I don't like the overall results of Tailwind's approach of condensing absolutely everything into tiny utility classes and applying all styling by adding an ever-increasing list of non-human-readable classes-with-modifiers onto each element.

Yes, it should be ugly, and maybe it's not perfect. But as we want to work more in a _component driven development_ way, once a component is customized, you don't have to edit any of the classes just use the component. Thus, code is written once.

_To be clear:_ 
I originally used to developed button like this: 
```html
<a class="button">
  Button
</a>
```
and yes, I really enjoyed _CSS framework_ that gives me this `button` class.

But now, in a _component driven development_ way, we use: 
```html
<Button>Button</Button>
```
and I really enjoyed having a highly customizable components like this:
```js
const Button = (label) => (
<button type="button" class="focus:outline-none text-white text-sm py-2.5 px-5 rounded-md bg-blue-500 hover:bg-blue-600 hover:shadow-lg">{label}</button>
)
```

**Comment by Matt-Yorkley** [comment]
> once a component is customized, you don't have to edit any of the classes just use the component.

That would be equally true if we leant on an existing library whilst making components, right?

> 99% of the styling is already handled by a library...

If we're working towards designs based on Material Design and we can either a) import something that's already done 99% of the work (and includes utility classes and SASS variables), or b) painstakingly build everything from scratch (starting from zero)... that's not a trivial thing, right?

> Thus, code is written once

Is it though? I was chatting to someone recently who worked with Tailwind and they ran into a conundrum of how to refactor their code. It looked like this: they had a set of styling applied to all H1 elements in the codebase. Every time they wanted to use a H1 title in a page, they had to replicate it, and ultimately it looked a bit like your button example, with a huge string of classes.

> class="focus:outline-none text-white text-sm py-2.5 px-5 rounded-md bg-blue-500 hover:bg-blue-600 hover:shadow-lg"

They had three options for how to deal with this:

1. Copy-paste this chunk of code for every page title. Obviously that's bonkers. It's apparently common for Tailwind devs though! 
2. Use Tailwind's `@apply` function to wrap that long list of classes into another class. This is a bit of an antipattern, it goes in a big circle and ends up back at the beginning with the exact same problems of specificity and inheritance that Tailwind is supposed to fix. It also requires a bunch of additional tooling added into the stack (which I think Tailwind requires in general).
3. Extract the code into it's own component. If a simple H1 tag is so messy it has to be extracted to a separate file just to use it, I feel like there's something going a bit wrong...

In the immortal words of [Gwen Stefani](https://www.youtube.com/watch?v=gZHjRQjbHrE); this shit is bananas. :banana::banana:

**Comment by Matt-Yorkley** [comment]
By "going round in a big circle and ending up at the beginning" I mean Tailwind starts by saying **this is bad**:
```html
<button class="btn btn-green">Click me!</button>
```
and **this is good**:
```html
<button class="py-2 px-4 font-semibold rounded-lg shadow-md text-white bg-green-500 hover:bg-green-700">Click me!</button>
```
But then when it comes to the issue of DRYing code, dealing with all that mess, making it re-usable etc, Tailwind ultimately [suggests doing this](https://tailwindcss.com/docs/utility-first#maintainability-concerns):
```css
  .btn {
    @apply py-2 px-4 font-semibold rounded-lg shadow-md;
  }
  .btn-green {
    @apply text-white bg-green-500 hover:bg-green-700;
  }
```
and then doing _this_ (wait, isn't this bad?!):
```html
<button class="btn btn-green">Click me!</button>
```
What? :joy: 

This is the main solution proposed by Tailwind in the [docs](https://tailwindcss.com/docs/utility-first#maintainability-concerns), but even Tailwind's creator has stated multiple times that it's an antipattern...

**Comment by jibees** [comment]
@Matt-Yorkley 
> That would be equally true if we leant on an existing library whilst making components, right?

Yes, if the existing library handles the customization of its _components_ (for Bootstrap, the class `btn` for example). As you said, seems to be done by MDBootstrap.

>> Thus, code is written once
> Is it though? 

At least, it's an objective. I know this can't be 100% applicable, but it's great to have such a thing in mind (and I know you have).

I don't know if we can reach an agreement if we both discuss and exchange arguments / counter-argument.

To step further, maybe we could: 
 - Invite more people to discuss here, or elsewhere?
 - and/or give a try to _your_ solution and _mine_ by creating one of the first components needed by Backoffice Product List such as #7199 inside the StyleGuide (#7284) with the two solutions. We can after compare them.



What do you think?

**Comment by jibees** [comment]
We are wondering if @openfoodfoundation/core-devs and @Erioldoesdesign (and others designers, UI/UX owners?) can add some thoughts about the **Tailwind or not** (or CSS utility first framework vs. _I don't exactly know how can i call them_ framework, but Material Design for Bootstrap seems to be the perfect example) **debate**. 
 
Thanks! 🙏

**Comment by mkllnk** [comment]
I learned a lot just by reading your posts and following a few links. And my conclusion is that you two know a lot more about CSS frameworks than me. :stuck_out_tongue: 

I'm not sure if the question here really is Tailwind vs MDBootstrap. Taking a step back, you expressed several needs:
- Plenty of theme work already done (e.g. a material-ish design).
- Easy to customise.
- We can build components with it.

What else do we need?

Once we know the needs, we can rate different options accordingly:

#### Tailwind

Using all the little classes can become messy. We can create new classes for our components. Why is that bad?

There may be a Tailwind theme that suits us so that we don't have to build everything from scratch.

The [tailwindcss-rails](https://github.com/rails/tailwindcss-rails) gem needs Rails 6. dhh created that gem. He's creating lots of good things.

My understanding is that Tailwind provides a large set of potentially useful building blocks and a framework to serve only the classes required. It seems flexible enough that we don't have to use it in a way that we don't like. Could it make our work more efficient?

#### MDBootstrap

I'm out of the loop with Bootstrap. It seems to have become a lot better and Matt says that it's a lot more customisable now.

**Comment by andrewpbrett** [comment]
I had a similar conclusion as @mkllnk :) 

I read a little about tailwind, and also had the urge to throw up a little bit when I saw all the tiny classes adding up on elements. But reading some other posts about it all follow the same pattern - there's a phase of disgust followed by enlightenment. So I wouldn't be terribly upset if we went with it. Plus if dhh is creating things for it... that's a pretty good indicator that there's **something** there that's pretty good.

**Comment by sauloperez** [comment]
Seriously, I LOVE TAILWIND :joy: you got to use it to understand it, and indeed you have an "aha! this makes total sense!" moment. It basically brings encapsulation and composition into the world of CSS.

However, it's also true that it favors full flexibility as opposed to having a set of components ready to be used (as most frameworks aim to provide). Nonetheless, I've been following the project and they've tried to overcome that drawback with things like https://tailwindui.com/ and https://headlessui.dev/ but still, we would still be on our own (these target JS frameworks). I understand our needs may require already made components.

**Comment by Erioldoesdesign** [comment]
Many of the UI Uplift designs *take inspiration from Material design* or may even be exactly as material design describes but my concern, which I've said before is that restricting ourselves to a framework that encourages or forces you to use their vanilla components means that we'll end up building custom components anyway *or* designs will be restricted when designing new features and improving older one to a sort of 'pain by numbers' or 'here are your building blocks' process.

For example, Mario used Material design chips in the new 'tagging list' UI: https://material.io/components/chips#usage

I mean he's changed the colour etc. but that's not hard to change in Material design so no biggie:
![Screenshot 2021-04-22 at 15 50 45](https://user-images.githubusercontent.com/11681324/115735656-8f4a0180-a382-11eb-9738-f3b146320e50.png)

So if we have a framework that allows us to use some material design elements we like, great! but what if Mario had a design idea for this tagging list that wasn't the material design 'chip' but a component not currently described in material design?

He had two options either:
1. Pick a piece of UI from material design that kind of works (not great for designers or users)
2. We create a new component in material design (extra dev time to create a new component in MD)

I think the decision is ultimately up to whoever will be maintaining the FE and working on the FE to choose what is the best choice of time and effort. I'm happy as long as design (and therefore the users experience) is restricted by us not being able to build custom components easily and pushing design to 'pick from the framework shelf' OFN is too much of a complex and detailed tool to get by on a 'simple' framework UI.

**Comment by jaycmb** [comment]
Does the Uplift take into account this one?
https://community.openfoodnetwork.org/t/quantities-on-hand-adaptating-when-variants-are-a-sub-product-of-the-main-product/480/4

I remember it being discussed but can´t recall anymore if that´s part of V1

**Comment by Erioldoesdesign** [comment]
@jaycmb Yes it was handled and changed to a different kind of UI and function when you click on the drop-down that is under the 'on hand' column heading:

![Screenshot 2021-05-10 at 10 19 54](https://user-images.githubusercontent.com/11681324/117637083-634bc000-b179-11eb-855e-f1f09c1ad0f9.png)

![Screenshot 2021-05-10 at 10 20 01](https://user-images.githubusercontent.com/11681324/117637088-63e45680-b179-11eb-9afd-585238eb297b.png)

![Screenshot 2021-05-10 at 10 20 09](https://user-images.githubusercontent.com/11681324/117637090-647ced00-b179-11eb-8912-529bef625972.png)

**Comment by RachL** [comment]
Wow! I always thought UI wasn't introducing any new features! And this particular one was always defined as complex, glad we could find an easy way to do it :muscle: :tada:

**Comment by RachL** [comment]
@Matt-Yorkley @lin-d-hop @filipefurtad0 and myself sat down in Portugal to refresh our memories on this feature and bring us up-to-date. While going through the issues, we reach agreements on some ideas that we are proposing here: 

1. Issues were made with components in mind, which is interesting but very difficult to pick up to understand what to deliver in the end. We would like to propose to **rewrite issues** in terms of user stories, with components as acceptance criteria. Rachel would be available to take on the creation of issues but would need to team up with a dev for the acceptance criteria.

While we rewrite issues, we need to keep in mind that some of them should be pick-up by back-end devs. So we need to make that more clear.

2. We need to be more clear on what are the expected results in terms of experience for the user. Some example of questions we had : 

- We should be careful about how the pagination works: infinite scroll or what we have now?

- How do filters behave : use one and get instant results ?

- Apparently we are introducing autosave! This should be shown to the user in some ways. Maybe a message at the top of the page? "All changes saved" ?

So it's linked to 1. but design decisions are needed.

3. The scope of the project is really wide, so we started to look at what would be the first minimal bits we want to ship in production. The proposal is the following : 

The first release should only focus on:
- removing Angular / introducing Tailwind (or whatever we choose)
- improving the current experience of the product page by introducing the new design. But the scope of features remains the same.

This means that we propose to do a first release without tags (which are not intended to work with existing tags in this project, so difficult to explain to user in the first release) and bulk actions on products. To be clear: we are not saying we want to remove these features from the project, just not targeting to have them in the first release / reveal of the new page to users.

Any thoughts @openfoodfoundation/core-devs @openfoodfoundation/train-drivers-product-owners @openfoodfoundation/testers ?

**Comment by kirstenalarsen** [comment]
Sounds great - let's get this show on the road!!

**Comment by RachL** [comment]
We had a little chat about the project today with @jibees 

**Summary**: 

We are still discovering new features hidden in the design! So again, for the purpose of scoping the first release to the bear minimum and ease dev work, the following changes will be made on the designs (I'm including stuff mentioned in my previous post + new stuff we discovered this morning):

1. Duplicate variants (only products can be duplicated)
2. You can't delete a variant when it's the only one left (only products can be deleted in that case)
3. Bulk actions on products will be removed, which means action button on top right + checkboxes
4. Tags are removed: filter options and column
5. We won't do infinite scrolling yet, so we keep pagination as it is today (only changing the UI / look and feel)
6. We do introduce autosave, so we need on the top left for example a little spinner that ends with the text "All changes saved"
7. Filters won't be used instantly, so we are introducing back the "filter results" button (easier on the dev side)

**Actions**: 

- Rachel to gather all existing designs to make sure we have the latest version (lots of links in the existing issues 😅 )
- JB to do a new version on figma based on the notes here
- Rachel to create issues user-oriented. We will start with an issue around displaying (read-only) products & the ability to filter them. Then actions buttons. Then the ability to edit on the page fields like name, unit, price, on hand and category (posssibly even one issue per fields).

**Comment by jibees** [comment]
I tried to :

 - add pagination
 - remove all references to tags
 - remove all references to bulk actions (including checkbox then)
 - add a "Filter Results" button
It's not perfect, but it's a first iteration that can be improve

<img width="1308" alt="Capture d’écran 2022-01-26 à 10 43 43" src="https://user-images.githubusercontent.com/296452/151140176-f88633ae-807f-405f-8b3c-fab1d8849a5e.png">

**Comment by kirstenalarsen** [comment]
Would it be scope creep with the pagination to have the option to ‘show X results’ that defaults to smaller number but lets them change it? Seems that’s pretty standard and if we’re setting precedent for admin pagination handling here would be good to include? 

I am sad to see tags go but I understand. I hope we can have them back soon because they are critical step towards removing Inventory

**Comment by jibees** [comment]
>Would it be scope creep with the pagination to have the option to ‘show X results’ that defaults to smaller number but lets them change it? Seems that’s pretty standard and if we’re setting precedent for admin pagination handling here would be good to include?

Right, we mention that during the chat we had, but I forgot to add in the design. Will do.

**Comment by jibees** [comment]
<img width="1039" alt="Capture d’écran 2022-01-27 à 10 21 53" src="https://user-images.githubusercontent.com/296452/151329704-86ff125e-942c-4b4c-8f84-5c409c8678dc.png">

**Comment by kirstenalarsen** [comment]
Thanks! Would it usually be down in the footer with the page numbers? I guess if setting it where you have it is more intuitive for them to be setting something that is saved as their default? (Like columns?)

**Comment by RachL** [comment]
Currently we have them on the top left of the table. They are not triggered by the "filter results" button. So what's the easiest on the dev side @jibees ? To make them work with the filter button or outside the button? 

>I am sad to see tags go but I understand. I hope we can have them back soon because they are critical step towards removing Inventory

Yep. It's because we agree on that that we've removed them actually. The way they were thought until now was just to add a new tag field NOT linked with the other tags in the system. Which - when discussing with the team - felt very weird for the user / difficult to explain. So the idea is to introduce them, but with a link with the existing tag system. I think that when we do this FR will not have any inventory user anymore :D

**Comment by RachL** [comment]
@jibees just noticing that your design is missing the action button per product, see: https://github.com/openfoodfoundation/openfoodnetwork/issues/8931

Also can we go up to 100 results per page? Or at least 50. 10 could be just one product and its variants, so it's tiny.

Let's have a call when you're back :)

**Comment by audez** [comment]
hi! I hope it's not too late to comment. First I really like the new UI, very nice, more intuitive, and the fact that all the variants are unfolded will be really precious for users!

The only thing that bothers me is that the "edit", "duplicate" or "delete" buttons will now be hidden under an "Actions" button. 
![Screen Shot 2022-09-09 at 15 17 42](https://user-images.githubusercontent.com/28386648/189359088-c461313f-bad2-4bdf-b454-3b1135204eb5.png)

I think it is very nice to have them "outside", as you have direct access to it, it's more visible and it spares one more click. And for the "duplicate" button, it is very useful as you can stay on the button and click several times on it and it will create several new products.
![Screen Shot 2022-09-09 at 15 20 36](https://user-images.githubusercontent.com/28386648/189359420-1b596716-d363-4e90-86e6-8aedff7a2754.png)

As the process of creating or editing products is already long, I really think it's a good think to have them outside, and keep the "Actions" button only for bulk actions. 

But maybe there's good reasons for this "Actions" button?

**Comment by dacook** [comment]
Before a full rollout, we should ensure that server monitoring is fully configured, so we can measure any impacts:
- https://github.com/openfoodfoundation/ofn-install/issues/922

( I wasn't able to add this to the milestone because it's in a different repo)

**Comment by RachL** [comment]
I don't think we need this epic anymore :tada:

**Comment by dacook** [comment]
It caused me to look back at the original goal of this epic, which was to encourage Inventory users to use the Bulk Products screen. It seems like tagging is intended to do that, and maybe scrolling laterally (side-scrolling).

**Comment by RachL** [comment]
@dacook yes we've never updated the epic content, tagging will be on of the next priorities ;) Regarding side scrolling, do you think you could do a rough estimate? I wonder if we could fund this, I think it would be very helpful.

**Comment by dacook** [comment]
We can do a 1hr spike, but I think it will require a few rounds to get right. It might require some re-structuring of the code too.
Also it's worth noting, it's probably not possible/practical for it to adapt to content (for example, if I have very long product names, the product column won't automatically expand). Maybe we would like to add column resizing but that's another story.

 A very quick roundup of requirements:
* First design:
https://www.figma.com/proto/kBvKYXnaNU0thSnKJUoyup/Product-List-UI-uplift?node-id=10-8745&viewport=116%2C270%2C0.04911578819155693&scaling=scale-down-width&hide-ui=1
* Second design: (I'm sure Mario set up a preview in the new design but can't find it right now)

When the viewport width is smaller, or a large number of columns is selected,
And the columns reach a minimum width (to be experimented) and no longer fit on the screen,
The remainder of the columns can be visible by horizontal scrolling to the right.

Considering:
* visual indicators to show when there's more content to be scrolled to (drop shadow on left/right)
* image and name columns to be sticky on the left
* actions menu to be sticky on the right?
* horizontal scroll bar should appear at bottom. It doesn't need to be sticky on the screen (there are ways to side-scroll with any mouse or touchpad)

I guess it depends how much we want to polish it.
----

Also, there's another enhancement we can do separately which will help for large screens. The Admin interface has a maximum width, but we can allow the table to break out of this:
![Screenshot 2024-09-03 at 10 15 27 am](https://github.com/user-attachments/assets/04e32fd9-32d5-41bf-8eec-9270c82f6b71)
